File: clients/python_pypy/codegame/client_message.py
from debugging.debug_command import DebugCommand
from model.order import Order
from stream_wrapper import StreamWrapper

class ClientMessage:
    """Message sent from client"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "ClientMessage":
        """Read ClientMessage from input stream
        """
        tag = stream.read_int()
        if tag == DebugMessage.TAG:
            return ClientMessage.DebugMessage.read_from(stream)
        if tag == OrderMessage.TAG:
            return ClientMessage.OrderMessage.read_from(stream)
        if tag == DebugUpdateDone.TAG:
            return ClientMessage.DebugUpdateDone.read_from(stream)
        if tag == RequestDebugState.TAG:
            return ClientMessage.RequestDebugState.read_from(stream)
        raise Exception("Unexpected tag value")

class DebugMessage(ClientMessage):
    """Ask app to perform new debug command"""

    TAG = 0

    __slots__ = ("command",)

    command: DebugCommand

    def __init__(self, command: DebugCommand):
        self.command = command
        """Command to perform"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "DebugMessage":
        """Read DebugMessage from input stream
        """
        command = DebugCommand.read_from(stream)
        return DebugMessage(command)
    
    def write_to(self, stream: StreamWrapper):
        """Write DebugMessage to output stream
        """
        stream.write_int(self.TAG)
        self.command.write_to(stream)
    
    def __repr__(self):
        return "DebugMessage(" + \
            repr(self.command) + \
            ")"

ClientMessage.DebugMessage = DebugMessage

class OrderMessage(ClientMessage):
    """Reply for ServerMessage::GetOrder"""

    TAG = 1

    __slots__ = ("order",)

    order: Order

    def __init__(self, order: Order):
        self.order = order
        """Player's order"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "OrderMessage":
        """Read OrderMessage from input stream
        """
        order = Order.read_from(stream)
        return OrderMessage(order)
    
    def write_to(self, stream: StreamWrapper):
        """Write OrderMessage to output stream
        """
        stream.write_int(self.TAG)
        self.order.write_to(stream)
    
    def __repr__(self):
        return "OrderMessage(" + \
            repr(self.order) + \
            ")"

ClientMessage.OrderMessage = OrderMessage

class DebugUpdateDone(ClientMessage):
    """Signifies finish of the debug update"""

    TAG = 2

    __slots__ = ()


    def __init__(self):
        pass

    @staticmethod
    def read_from(stream: StreamWrapper) -> "DebugUpdateDone":
        """Read DebugUpdateDone from input stream
        """
        return DebugUpdateDone()
    
    def write_to(self, stream: StreamWrapper):
        """Write DebugUpdateDone to output stream
        """
        stream.write_int(self.TAG)
    
    def __repr__(self):
        return "DebugUpdateDone(" + \
            ")"

ClientMessage.DebugUpdateDone = DebugUpdateDone

class RequestDebugState(ClientMessage):
    """Request debug state from the app"""

    TAG = 3

    __slots__ = ()


    def __init__(self):
        pass

    @staticmethod
    def read_from(stream: StreamWrapper) -> "RequestDebugState":
        """Read RequestDebugState from input stream
        """
        return RequestDebugState()
    
    def write_to(self, stream: StreamWrapper):
        """Write RequestDebugState to output stream
        """
        stream.write_int(self.TAG)
    
    def __repr__(self):
        return "RequestDebugState(" + \
            ")"

ClientMessage.RequestDebugState = RequestDebugState

File: clients/kotlin_graal/src/main/kotlin/ai_cup_22/codegame/ClientMessage.kt
package ai_cup_22.codegame

import ai_cup_22.util.StreamUtil

/**
 * Message sent from client
 */
abstract class ClientMessage {
    /**
     * Write ClientMessage to output stream
     */
    @Throws(java.io.IOException::class)
    abstract fun writeTo(stream: java.io.OutputStream)

    companion object {
        /**
         * Read ClientMessage from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): ClientMessage {
            when (StreamUtil.readInt(stream)) {
                DebugMessage.TAG -> return DebugMessage.readFrom(stream)
                OrderMessage.TAG -> return OrderMessage.readFrom(stream)
                DebugUpdateDone.TAG -> return DebugUpdateDone.readFrom(stream)
                RequestDebugState.TAG -> return RequestDebugState.readFrom(stream)
                else -> throw java.io.IOException("Unexpected tag value")
            }
        }
    }

    /**
     * Ask app to perform new debug command
     */
    class DebugMessage : ClientMessage {
        /**
         * Command to perform
         */
        var command: ai_cup_22.debugging.DebugCommand
    
        constructor(command: ai_cup_22.debugging.DebugCommand) {
            this.command = command
        }
    
        /**
         * Write DebugMessage to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            command.writeTo(stream)
        }
    
        /**
         * Get string representation of DebugMessage
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("DebugMessage { ")
            stringBuilder.append("command: ")
            stringBuilder.append(command)
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 0
    
            /**
             * Read DebugMessage from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): DebugMessage {
                var command: ai_cup_22.debugging.DebugCommand
                command = ai_cup_22.debugging.DebugCommand.readFrom(stream)
                return DebugMessage(command)
            }
        }
    }

    /**
     * Reply for ServerMessage::GetOrder
     */
    class OrderMessage : ClientMessage {
        /**
         * Player's order
         */
        var order: ai_cup_22.model.Order
    
        constructor(order: ai_cup_22.model.Order) {
            this.order = order
        }
    
        /**
         * Write OrderMessage to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            order.writeTo(stream)
        }
    
        /**
         * Get string representation of OrderMessage
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("OrderMessage { ")
            stringBuilder.append("order: ")
            stringBuilder.append(order)
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 1
    
            /**
             * Read OrderMessage from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): OrderMessage {
                var order: ai_cup_22.model.Order
                order = ai_cup_22.model.Order.readFrom(stream)
                return OrderMessage(order)
            }
        }
    }

    /**
     * Signifies finish of the debug update
     */
    class DebugUpdateDone : ClientMessage {
    
        constructor() {
        }
    
        /**
         * Write DebugUpdateDone to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
        }
    
        /**
         * Get string representation of DebugUpdateDone
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("DebugUpdateDone { ")
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 2
    
            /**
             * Read DebugUpdateDone from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): DebugUpdateDone {
                return DebugUpdateDone()
            }
        }
    }

    /**
     * Request debug state from the app
     */
    class RequestDebugState : ClientMessage {
    
        constructor() {
        }
    
        /**
         * Write RequestDebugState to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
        }
    
        /**
         * Get string representation of RequestDebugState
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("RequestDebugState { ")
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 3
    
            /**
             * Read RequestDebugState from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): RequestDebugState {
                return RequestDebugState()
            }
        }
    }
}